plugins/rest/rest.go

```diff
@@ -91,10 +98,11 @@ func (c *Config) authPrepare(req *http.Request) error {
 // Client implements an HTTP/REST client for communicating with remote
 // services.
 type Client struct {
-	bytes   *[]byte
-	json    *interface{}
-	config  Config
-	headers map[string]string
+	bytes            *[]byte
+	json             *interface{}
+	config           Config
+	headers          map[string]string
+	authPluginLookup func(string) HTTPAuthPlugin
```

Comment:
Is there a reason we don't invoke the function when we're creating the `Config` in the `rest.New` function? This way we don't have to pass this function around (e.g., as a paremeter to `authPrepare`, `authHTTPClient`, etc.)

Reply:
Hmm, that's a good question. I was simply following the pattern the other auth plugins use, but you may be right that it makes more sense to just setup the auth plugin (any of them) when creating the rest client.  I need to dig a bit more on this, but maybe @nhw76 or @patrick-east could chime in?  It looks like they touched this code last.

Reply:
It sure does seem like a new auth client is being created for every single HTTP request. That's gotta be super inefficient, although changing that behavior could be a breaking change, because a side effect of recreating the auth plugin for each request is that rotated credentials are automatically used, e.g. mutual tls certs are read from disk for each request.

Assuming I understand the current code, I'm wary of refactoring that myself, especially if I have to then go add code to deal with credentials refreshing to handle backwards compatibility.

Any chance we can keep the authPluginLookup passing as is for this PR (with your other feedback addressed), and attempt to refactor the auth plugins in a separate one?

Reply:
> although changing that behavior could be a breaking change, because a side effect of recreating the auth plugin for each request is that rotated credentials are automatically used, e.g. mutual tls certs are read from disk for each request because clientTLSAuthPlugin.NewClient is called in rest.Do.

Technically, I think you're correct. However, it may not be a big issue, when looking at the concrete available plugins:

- `clientTLSAuthPlugin`: I don't know if people really change TLS certs deployed with a service, and expect the service to pick up changes to them, using the latest TLS certs from the filesystem. Personally, I'd be surprised by a service doing that 😄 
- `gcpMetadataAuthPlugin` gets a token from the metadata service, and doesn't deal with refresh tokens of any sort. 
- `bearerAuthPlugin` is entirely static, `NewClient` per request or per-startup doesn't matter.
- `oauth2ClientCredentialsAuthPlugin` can get a fresh token from a call to `Prepare`
- `aws` does what it needs to do from `Prepare`, too, if I understand the code correctly.

So, having a client per-startup vs per-request perhaps only makes a difference for the TLS case, where I'd think that's acceptable. 🤔

Reply:
> Any chance we can keep the authPluginLookup passing as is for this PR (with your other feedback addressed), and attempt to refactor the auth plugins in a separate one?

I don't think the inefficiency of recreating the clients is an issue here (it's unlikely to be a bottleneck and the immutable nature of the clients is helpful in avoiding state and keeping things simple.) That said, the code around the clients has gotten a bit complicated so I understand not wanting to touch it. If you find it's difficult to just move the `authLookup` call into `rest.New` feel free to leave it as is.